model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanAuthenticationSessionProviderFactory.java

```diff
@@ -78,25 +92,66 @@ public static int getAuthSessionsLimit(Config.Scope config) {
         return limit <= 0 ? DEFAULT_AUTH_SESSIONS_LIMIT : limit;
     }
 
-
     @Override
     public void postInit(KeycloakSessionFactory factory) {
+        this.remoteCacheInvoker = new RemoteCacheInvoker();
+        keyGenerator = new InfinispanKeyGenerator();
         factory.register(new ProviderEventListener() {
 
             @Override
             public void onEvent(ProviderEvent event) {
                 if (event instanceof PostMigrationEvent) {
-
                     KeycloakModelUtils.runJobInTransaction(factory, (KeycloakSession session) -> {
-
                         registerClusterListeners(session);
-
+                        loadAuthSessionsFromRemoteCaches(session);
                     });
                 }
             }
         });
     }
 
+    private void loadAuthSessionsFromRemoteCaches(KeycloakSession session) {
+        for (String cacheName : remoteCacheInvoker.getRemoteCacheNames()) {
+            loadAuthSessionsFromRemoteCaches(session.getKeycloakSessionFactory(), cacheName, getSessionsPerSegment(), getMaxErrors());
+        }
+    }
+
+    private int getStalledTimeoutInSeconds(int defaultTimeout) {
+        return config.getInt("stalledTimeoutInSeconds", defaultTimeout);
+    }
+
+    private void loadAuthSessionsFromRemoteCaches(final KeycloakSessionFactory sessionFactory, String cacheName, final int sessionsPerSegment, final int maxErrors) {
```

Comment:
@ahus1 @mhajas can you get your feedback on this? We deprecated the remote store and the old logic didn't fetch the data from the remote store. Do you think it makes sense to add this now?

Reply:
@pruivo @graziang I might miss some context and don't have latest knowledge of the infinispan/cluster/cross-dc layer in general, but AFAIK we never supported backup of `authenticationSessions` cache by the remote-store in the cross-dc environment in the old store (Infinispan distributed caches). We supported it for caches like `sessions`, `clientSessions`, `offlineSessions`, `loginFailures` etc, but not for `authenticationSessions` . So my vote is to not add it as part of this PR.

If I understand correctly, the related issues, which this PR solves ( #23881 and #32658 ) does not need `authenticationSessions` cache backed by the remote-store. So not needed to include in this PR IMO.

Reply:
Also, I suspect it is broken since if it imports from an older version, the remote store will have `RootAuthenticationSessionEntity` stored. It will throw a `ClassCastException` because we store `SessionEntityWrapper` in the embedded cache.

Reply:
> AFAIK we never supported backup of authenticationSessions cache by the remote-store in the cross-dc environment in the old store (Infinispan distributed caches).

Hmm this is interesting. It really seems initially, in tech preview of crossdc (kc 18), we didn't support/configure remote store for authentication sessions: https://github.com/keycloak/keycloak/blob/18.0.2/testsuite/integration-arquillian/servers/auth-server/jboss/common/jboss-cli/cross-dc-setup.cli but then with the first version of multi-site in Keycloak 24 we added it because we assumed it is necessary :/ https://github.com/keycloak/keycloak-benchmark/blob/6eb3efdb2007c5fa2b03223379e6d1bca20cdec9/provision/minikube/keycloak/config/kcb-infinispan-cache-remote-store-config.xml#L70-L96

Since we added support for it, I am afraid we will need to keep it part of this PR since we cannot do any breaking changes in minor release. 


> Also, I suspect it is broken since if it imports from an older version, the remote store will have RootAuthenticationSessionEntity stored. It will throw a ClassCastException because we store SessionEntityWrapper in the embedded cache.

@pruivo What do you mean by this? We are not changing the protostream representation here, so it should not be broken.

Reply:
I am wondering, if it would be simpler to create a new AuthenticationSession provider without remote store support with fixes https://github.com/keycloak/keycloak/issues/23881 and https://github.com/keycloak/keycloak/issues/32658 and hide it behind a disabled_by_default feature and switch it to default in next major. WDYT?

Reply:
You still need to do some migration at some point to convert from one class to another.

Reply:
Yes, but we can do it in the next major version while keeping the previous behavior for minor releases

Reply:
@mhajas so, do you want to delay this PR until Keycloak 27?

Keycloak 26.0.x has the remote store deprecated and stores `RootAuthenticationSessionEntity`. 
For Kecyloak 26.1.x (assuming the remote store is kept deprecated), if you want to merge this PR, you need to fetch the data from the remote store and replace/wrap it with the `SessionEntityWrapper`.

Am I wrong?

Reply:
> @mhajas so, do you want to delay this PR until Keycloak 27?

No. What I am saying is that we now support storing authentication sessions in the remote store, therefore, we cannot remove support for it in this PR. This brings a lot of work needed to make this work with the ChangeLogBasedTransaction used for sessions as we are doing a lot of ugly things, like skipping of remote store etc. If I understand correctly for this reason we are also preloading the authentication sessions from remote store for this reason.

My suggestion is to introduce a new provider that would not have support for remote store at all that would not have issues https://github.com/keycloak/keycloak/issues/23881 and https://github.com/keycloak/keycloak/issues/32658. This new provider would be disabled by default and could be enabled by a feature like AUTHENTICATION_SESSIONS:v2 because it is breaking change as it does not support remote store. 

In Keycloak 27, we would switch it to be enabled by default and the old provider would be removed with the CACHE_EMBEDDED_REMOTE_STORE feature removed.


Does it make more sense now?

Reply:
Users using the deprecated code will not have the issue fixed with that approach. We would have to duplicate the tests to enable the feature and ensure the tests are passing with the new implementation.

The number of issues I answered because users trying to do a rolling upgrade between KC25 -> KC26, and ProtoStream was complaining led me to believe they will never enable it :laughing:

Reply:
> Users using the deprecated code will not have the issue fixed with that approach.

This issue is there probably since the beginning of Keycloak so I don't think it is that urgent to fix it asap.

> We would have to duplicate the tests to enable the feature and ensure the tests are passing with the new implementation.

That is true. However, we will need to find some good solution for this as we plan to utilize features more in future.

> The number of issues I answered because users trying to do a rolling upgrade between KC25 -> KC26, and ProtoStream was complaining led me to believe they will never enable it 😆

Then they will need to wait for fixing of the issues for the next major release. If they want the fix, they have the possibility to enable the feature.



We have 3 scenarios where authentication sessions are used. 
1. Single site using only embedded ISPN
2. Multi-site using the new Remote* implementation
3. Deprecated CACHE_EMBEDDED_REMOTE_STORE

This PR is trying to fix the bug in all 3. My assumption is that if we don't consider fixing it in 3 we could make it simpler and we could avoid some unexpected behavior, for example, when `InfinispanChangelogBasedTransaction` uses some unexpected `SKIP_CACHE_LOAD` flag to some operation that `AuthenticationSessionProvider` does not expect and similar.

But I must admit I didn't go through the PR thoroughly so maybe I am missing some other aspect of this. I am not saying my suggestion is the best option. What I am saying is that if we want to save some work to make remote-store work, there is an option. @pruivo If you want we can jump on a call and discuss it there.

Reply:
I don't want to fix it for `CACHE_EMBEDDED_REMOTE_STORE` either, so importing it from the remote store is not worth it; we can just clear the remote store using a migration method. 

`InfinispanChangelogBasedTransaction` only decorates the cache for writes, where it writes to the embedded caches first and then to the external cluster. I'm optimistic here so I don't expect any issues. (although this is not efficient at all)

You know my desire of rewriting the embedded providers from scratch :smile:

Reply:
Ok, I thought we are using `SKIP_CACHE_LOAD` also in the transaction while it is used only in the `UserSessionProvider`. Ok then. We can proceed with what we have. Cleaning remote store on migration sounds correct.